odoo/addons/base/models/res_lang.py

```diff
@@ -155,7 +170,223 @@ def fix_datetime_format(format):
             tools.resetlocale()
 
     @api.model
-    def install_lang(self):
+    def _get_i18n_url(self, base, lang):
+        """ Generate the URL to fetch the translation resource
+            e.g. https://nightly.odoo.com/i18n/13.0/fr.tar.xz
+        """
+        version = release.version
+        return urljoin(base, "i18n/{}/{}.tar.xz".format(version, lang))
+
+    @api.model
+    def _read_po_from_i18n_folder(self, addons, lang):
+        path = get_module_resource(addons, 'i18n', lang + '.po')
+        if path:
+            with open(path, 'r') as f:
+                yield f
+
+        extra_path = get_module_resource(addons, 'i18n_extra', lang + '.po')
+        if extra_path:
+            with open(extra_path, 'r') as f:
+                yield f
+
+        if not path and not extra_path:
+            _logger.info('module %s: no translation for language %s', addons, lang)
+
+    @api.model
+    def _read_po_from_attachment(self, addons, lang):
+        module = self.env['ir.module.module'].search([('name', '=', addons)])
+        if not module:
+            return None
+
+        attachment_name = "{}/i18n/{}.po".format(addons, lang)
+        po_file = self.env['ir.attachment'].search([
+            ('res_model', '=', 'ir.module.module'),
+            ('res_id', '=', module.id),
+            ('name', '=', attachment_name)
+        ], limit=1)
+
+        if po_file:
+            yield StringIO(po_file.datas)
+
+    @api.model
+    def _extract_po_to_i18n_folder(self, addons, tmp, extracted_file):
+        """ Extraction method if the po are stored on filesystem, inside /i18n/ folders """
+        path = get_module_path(addons)
+        if not os.access(path, os.W_OK):
+            _logger.warning("Path %s is not writeable", path)
+            return False
+
+        src_file = os.path.join(tmp, extracted_file)
+        dst_file = os.path.join(path, extracted_file)
+        shutil.move(src_file, dst_file)
```

Comment:
Is the `addons_path/addon/i18n` directory guaranteed to exist? If not,  I believe the move will fail if it does not exist.

Reply:
@naglis the path not existing is a subset of the path not being writable, so `os.access(path, os.W_OK)` will return False if the path doesn't exist.